tools/compactsnoop.py

```diff
@@ -0,0 +1,389 @@
+    data.nid = valp->nid;
+    data.idx = valp->idx;
+    data.order = valp->order;
+    data.sync = valp->sync;
+
+#ifdef EXTNEDED_FIELDS
+    data.fragindex = valp->fragindex;
+    data.min = valp->min;
+    data.low = valp->low;
+    data.high = valp->high;
+    data.free = valp->free;
+#endif
+
+    data.status = status;
+    data.stack_id = stack_traces.get_stackid(ctx, 0);
+
+    events.perf_submit(ctx, &data, sizeof(data));
+
+    start.delete(&id);
+}
+
+#ifdef EXTNEDED_FIELDS
+int trace_fragmentation_index_return(struct pt_regs *ctx)
+{
+    struct val_t val = { };
+    int ret = PT_REGS_RC(ctx);
+    u64 id = bpf_get_current_pid_tgid();
+    val.fragindex = ret;
+    start.update(&id, &val);
+    return 0;
+}
+#endif
+
+static inline void fill_compact_info(struct val_t *valp,
+                                     struct zone *zone,
+                                     int order)
+{
+    valp->nid = zone_to_nid_(zone);
+    valp->idx = zone_idx_(zone);
+    valp->order = order;
+}
+
+RAW_TRACEPOINT_PROBE(mm_compaction_suitable)
+{
+    // TP_PROTO(struct zone *zone, int order, int ret)
+    struct zone *zone = (struct zone *)ctx->args[0];
+    int order = (int)ctx->args[1];
+    int ret = (int)ctx->args[2];
+    u64 id;
+
+    if(ret != COMPACT_CONTINUE)
+        return 0;
+
+    id = bpf_get_current_pid_tgid();
+
+#ifdef EXTNEDED_FIELDS
+    struct val_t *valp = start.lookup(&id);
+    if (valp == NULL) {
+        // missed entry or order <= PAGE_ALLOC_COSTLY_ORDER, eg:
+        // manual trigger echo 1 > /proc/sys/vm/compact_memory
+        struct val_t val = { .fragindex = -1000 };
+        valp = &val;
+        start.update(&id, valp);
+    }
+    fill_compact_info(valp, zone, order);
+    get_all_wmark_pages(zone, valp);
+#else
+    struct val_t val = { };
+    fill_compact_info(&val, zone, order);
+    start.update(&id, &val);
+#endif
+
+    return 0;
+}
+
+RAW_TRACEPOINT_PROBE(mm_compaction_begin)
+{
+    // TP_PROTO(unsigned long zone_start, unsigned long migrate_pfn,
+    //          unsigned long free_pfn, unsigned long zone_end, bool sync)
+    bool sync = (bool)ctx->args[4];
+
+    u64 id = bpf_get_current_pid_tgid();
+    struct val_t *valp = start.lookup(&id);
+    if (valp == NULL) {
+        // missed entry
+        return 0;
+    }
+
+    valp->ts = bpf_ktime_get_ns();
+    valp->sync = sync;
+    return 0;
+}
+
+RAW_TRACEPOINT_PROBE(mm_compaction_end)
+{
+    // TP_PROTO(unsigned long zone_start, unsigned long migrate_pfn,
+    //          unsigned long free_pfn, unsigned long zone_end, bool sync,
+    //          int status)
+    submit_event(ctx, ctx->args[5]);
+    return 0;
+}
+"""
+
+if args.extended_fields:
+    bpf_text = EXTENDED + bpf_text
+else:
+    bpf_text = NO_EXTENDED + bpf_text
+
+if args.pid:
+    bpf_text = bpf_text.replace("PID_FILTER",
+                                "if (pid != %s) { return 0; }" % args.pid)
+else:
+    bpf_text = bpf_text.replace("PID_FILTER", "")
```

Comment:
I did not seee PID_FILTER in the above program.

Reply:
Sorry for this, I'll add it.